### PR TITLE
New update mechanism

### LINKED ISSUE #36: Command line attribute `rev` is ignored
The rev attribute is ignored when provided on CLI.

Considering `niv add NixOS/nixpkgs -a rev=plop`, the `sources.json` should contains `rev=plop` while rev have been updated according to the `--branch` attribute.

Specifying a `rev` attribute is useful when switching a project to Niv because it allows to introduce Niv without changing what is actually built.

### PR DESCRIPTION
**Pull request by nmattia**
Fixes #36 .

Got a few things to clean up, but this will greatly facilitate the move to multiple fetchers, and fixes a few bugs.

* Introduces the `Update`, which takes care of lazy loading the IO actions (not run if never read or if value will eventually be discarded)
* Adds a new (unit) test infrastructure
* Moves from `String` to `Text`
* A bit of renaming

Few things left to iron out:
* Re-gen the silly svg
* Check the error handling
* Remove all occurrences of `foo`, `bar`, `baz`, etc
* Clarify the separation between the "CLI" spec and the existing spec (i.e. clarify by changing the variable names)
* Document the hell out of everything

CC @nlewo @michaelpj

### CODE REVIEW

**Comment by nmattia** [comment]
Note to self: it'd be better to discarded the `null` values before serializing

File: script/test [hunk 1]
```diff
@@ -11,6 +11,6 @@ export NIX_PATH="nixpkgs=./nix"
 echo "Building"
 
 # Build and create a root
-nix-build --no-link
+nix-build --sandbox --no-link --max-jobs 10
```
Comment (nlewo) [inline script/test]: Does it work with the default Nix installation (sanboxing is not available)?

Comment (nmattia) [reply script/test]: What do you mean default Nix installation? On both Linux and macOS sandbox works for me by using the default installer (and nothing more) (albeit with some tweaks on macOS).

We do want to test that everything works with the sandbox; if someone doesn't have support for the sandbox they can simply run `nix-build`.

What do you think?

Comment (nlewo) [reply script/test]: I though sandbox was not working by using the default installer. So, if it is working by default, that's fine with me.

**Pull request merged**